TEIIDSB-126 [simpledb] Adding Amazon SimpleDB Source

ciphereck opened the pull request:
Signed-off-by: Mudit Jain <ciphereck@gmail.com>

ciphereck commented:
@shawkins 
Can you review the PR once? If the code looks good? Are there any tests required?
I wrote this in accord with other modules like S3 or Cassandra.

ciphereck commented:
I think it's failing since teiid-parent latest snapshot vs 15 which is not synced with the latest code of master of teiid-parent.

ciphereck commented:
@rareddy can you review it once?

shawkins commented on data/amazon-simpleDB/src/main/java/com/teiid/spring/data/amazon/simpleDB/SimpleDBConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+    private String accessKey;
+    private String secretAccessKey;
+
+    public SimpleDBConfiguration(String accessKey, String secretAccessKey) {
```
The constructor won't need these as arguments.

ciphereck replied:
So, it's going to be a default constructor with no arguments? And keys will only be set by setters?

If yes? why
if no? then what else?
I'm confused here?

aditya300899 replied:
I think the compiler automatically provides a default constructor, so there won't be a need to create that explicitly. The data sources are configured with the help of the application.properties file in SpringBoot by creating a bean for the configuration class. See this https://github.com/teiid/teiid-spring-boot/blob/master/samples/hdfs/src/main/java/org/teiid/spring/example/DataSources.java. I hope I've made myself clear. Otherwise, @shawkins will help.

ciphereck replied:
Thanks Aditya for the explanation. Yeah, no need to write default constructor explicitly, that's for sure. 
But I think I get what you are trying to say. So, I guess it's like instead of creating an object using argument constructor explicitly, legacy code uses setters to set the value of various parameters?

aditya300899 replied:
Yes, it's a feature of SpringBoot known as Externalized configuration.

ciphereck replied:
Sounds good then. Thanks for the clarification.

rareddy replied:
In spring boot the configuration works through introspection/injection, as in the framework bootstraps the object and injects the configured data before giving it to the running program. So, there if you use custom constructors then spring boot will not able to instantiate the constructor, this problem does not exist with default constructors with getters and setter methods. In general, whenever you want the object to be un-mutable with some state then you should consider the constructor based parameters and not provide any setters, then no one changes the state.

shawkins commented:
> I think it's failing since it used teiid-parent latest snapshot version 15 which is not synced with the latest code of master of teiid-parent.

I'll update the snapshot.

ciphereck commented:
> I'll update the snapshot.

Can you please leave a comment here, after updating the snapshot.

shawkins commented:
> Can you please leave a comment here, after updating the snapshot.

My build unfortunately failed.  I'll try running it again, and travis should produce a snapshot today as well.

Other things that are needed:  
- adding the spring data sample for simpledb.  Running that should also have highlighted springs usage of the configuration class
- a mustache file to automate the creation of the spring datasource class for usage with the operator
- add the module as a dependency to spring data util as we currently require a generated file that this module produces to make the sources known to operator

ciphereck commented:
@shawkins 
The build is passing now. I think the snapshot is in sync with our requirement.

I'm starting my work on spring data sample, on another branch. I'll raise another PR for the same.
Meanwhile, if everything looks good to you on this PR, can we move towards merging it?

ciphereck commented:
I'm working on installing SimpleDB locally, so that I can test it with module example-simpledb
I'll update here as soon if I get any.

rareddy commented on data/amazon-simpleDB/src/main/java/com/teiid/spring/data/amazon/simpleDB/SimpleDBConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+
+public class SimpleDBConfiguration implements BaseSimpleDBConfiguration {
+    private String accessKey;
+    private String secretAccessKey;
```
`secretAccessKey` is password? this variable seems odd? How it called on SimpleDB?

ciphereck replied:
It's actually secretKey of aws.
Originially in the wildfly module, this was the variable name throughtout, that's why this. 

https://github.com/ciphereck/teiid/blob/d3737491eae3ad49b9a6d41c8452d0e969cf092a/wildfly/connectors/amazon/connector-simpledb/src/main/java/org/teiid/resource/adapter/simpledb/SimpleDBConnectionImpl.java#L48

I didn't understood "How it called on SimpleDB?" part?

rareddy replied:
check with amazon-athena, or s3 nomenclature for properties see if you can map to that in spring-boot

ciphereck replied:
Well in s3, it's secretKey only.
But to do mapping, we need to change interface in teiid-parent too.

Since, in last SimpleDBConnectionImpl, it was secretAccessKey, I took the same as forward.

shawkins replied:
I'm fine with changing this on the teiid side to just secretKey.  We can leave the resource adapter with the secretAcessKey for backwards compatibility.

ciphereck replied:
I didn't get it. Here when you say to teiid side, do you mean teiid-parent or teiid-spring-boot? Cause you want me to leave resource adapter (I think it means connector in teiid-parent).

Anyways this is what I came up with: [ciphereck@652a44](https://github.com/ciphereck/teiid/commit/652a449)
I'm not sure, whether changing this, will trigger any issue or not. If this seems like a good idea, I can create a PR for the same. Otherwise, please let me know how to proceed.

rareddy replied:
`teiid-parent` project is the main `teiid` project, and then `teiid-spring-boot` is dependent on it. but both are separate projects, teiid-spring-boot depends on teiid but not other way around, so when you write code you need to follow that.

ciphereck replied:
Yeah. But to change the getter from getSecretAccessKey() to getSecretKey(), we need to change interface it in the teiid-parent simpledb module.

rareddy replied:
yes, go ahead and change there and submit a PR, but as Steve says do not change in the resource-adapter part as it will break the backward compatibility. so resource-adapter's SimpleDbConfiguration class you will have the new method and old but old method will be used as set/get and new method just uses values from old one.

ciphereck replied:
If we use old method as set/get only, will it situation, we are facing currently?
What I could understand is we now have two method, old one which return secretAccessKey and a new default method in interface which return value from old one. I think this is the only change we need to incorporate it in spring-boot.

Am I right here?

ciphereck replied:
teiid/teiidt#1311

rareddy replied:
In spring-boot ignore the old method and only define getters and setters for new one. To support the interface you can have the getter for the old but pointing to the new method

rareddy commented on data/pom.xml:
```diff
@@ -34,6 +34,7 @@
         <module>hdfs</module>
         <module>actian</module>
         <module>amazon-athena</module>
+        <module>amazon-simpleDB</module>
```
make it like `amazon-simpledb` loose capital `DB`, it also means changing the directory name with similar semantics

ciphereck replied:
yeah this seems right. Let me make another commit here quickly.

ciphereck commented:
Added Mustache file. Took inspiration from cassandra and s3 module and wrote it intuitively. Not sure, what's the standard way of write it. I request to check it once throughly.

Changed the secetAccessKey to secretKey and it's setter along with it. 
But couldn't change the getter, cause it's implemented from interface which is in teiid-parent. Seems odd, because the setters and getters name are in different way. But this is the immediate fix.

A more correct solution can be to first update the interface in teiid-parent and refactor it accordingly, update the snapshot and then update it in teiid-spring-boot. If you want, I can quickly create a fix for this in teiid-parent.

rareddy commented on data/amazon-simpledb/src/main/java/com/teiid/spring/data/amazon/simpledb/SimpleDBConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+    public String getSecretAccessKey() {
+        return secretKey;
+    }
+
```
Needs a getSecretKey() method here.

ciphereck replied:
Since, there was a default method, we don't need to implement it in SimpleDBConfiguration class.
It will automatically call getSecretAccessKey whenever we call getSecretKey method.
So, it's not required.

rareddy replied:
I think last time I did that the spring introspection did not behave well, as it looks for bot get and set methods and did not find one from default I belive. Also see the latest change that @shawkins did on Teiid side on this too

ciphereck replied:
@rareddy I updated the code, and the build is now passing. Can we move towards closing the PR?

cc: @shawkins

ciphereck commented:
We need to update the snapshot, for the build to pass. It's building on my local laptop with success.

shawkins commented:
> We need to update the snapshot, for the build to pass. It's building on my local laptop with success.

I'll update the snapshot.  I'll update travis to just create one on every build.

ciphereck commented:
1. I was thinking, if can we also update travis to send build result of failure and success on email, once the build either completes or fails.
2. I was also thinking of the same, it will help for future development as well. I'm curious how you update the snapshot by the way.

ciphereck commented:
> Travis does notify of failure or when a build starts passing again. Do you have a travis account?

I don't have a travis account. Just making one. Btw what I meant was to notify build status on email based on PR participants even if they don't have a travis account or at least to author.


> Also the travis settings are integrated to gitter: https://github.com/teiid/teiid/blob/0f06b7d07f7782bdbdb4c63662a4cc22898d63ea/.travis.yml#L25
> 
> So a failure will always show up in the chat log, but success only when it was previously failing.

I'm not able to find a channel of gitter here? I mean I'm not able to join or open it.

> I'm not sure what you are looking for in terms of the parallel development other than having the teiid snapshot built with each change.

Every change is on master I guess?

shawkins commented:
> Btw what I meant was to notify build status on email based on PR participants even if they don't have a travis account or at least to author.

See https://docs.travis-ci.com/user/notifications/#:~:text=Travis%20CI%20only%20sends%20build,the%20Travis%20CI%20notification%20settings.

Seems to imply that you need a travis account and have github email verified.  You could do some digging to see if requiring a travis account is optional.

> I'm not able to find a channel of gitter here? I mean I'm not able to join or open it.

Try https://gitter.im/teiid/GSOC by default the https://gitter.im/teiid room is restricted to members of the teiid org.

I've added the webhook that should show all github events in that channel.  The travis events only specify the webhook for https://gitter.im/teiid though.

> Every change is on master I guess?

Every change that would affect the state of the snapshot that is referenced by Teiid Spring Boot master yes.  You can of course work on both projects in your own branches.

> Btw when you update the snapshot, can you update it once here? Also, how to refresh travis build?

You mean how can you restart a travis build?  Once you have a travis account, you navigate to the pr build https://travis-ci.org/github/teiid/teiid-spring-boot/builds/709227659 and there is a restart build link on the page.  Otherwise any git change will trigger another build.

ciphereck commented:
> You mean how can you restart a travis build? Once you have a travis account, you navigate to the pr build https://travis-ci.org/github/teiid/teiid-spring-boot/builds/709227659 and there is a restart build link on the page. Otherwise any git change will trigger another build.

I'm not getting any restart button. But I'll figure out something. I checked stackoverflow, only admins can restart the build using travis-ci. Thanks for the detail explanations btw. 

<img width="1451" alt="Screenshot 2020-07-18 at 10 56 39 AM" src="https://user-images.githubusercontent.com/24265219/87845595-5e660900-c8e6-11ea-9998-e6fdb5357efb.png">

shawkins commented on pom.xml:
```diff
@@ -844,6 +855,11 @@
+      <dependency>
+        <groupId>org.teiid</groupId>
+        <artifactId>spring-data-amazon-simpledb</artifactId>
+        <version>1.6.0-SNAPSHOT</version>
```
One additional small change.  This should reference ${project.version} instead, I've updated that on master.

ciphereck replied:
Ohh yes. Idk how this got left by me. Thanks for finding and updating it.
